store/sqlstore/channel_member_history_store.go

```diff
@@ -119,3 +185,8 @@ func (s SqlChannelMemberHistoryStore) PermanentDeleteBatch(endTime int64, limit
 		}
 	})
 }
+
+// used by tests to truncate data in the store
+func (s SqlChannelMemberHistoryStore) GetDbMaster() *gorp.DbMap {
+	return s.GetMaster()
```

Comment:
Is this necessary? I think a type assertion would be cleaner as you wouldn't have to add a SQL-specific interface method.

Reply:
What do mean by a type assertion? I don't see any other way to get an instance of `gorp.DbMap` in the test than to add this method to the interface. When in the tests, we're dealing with the interface type `ChannelMemberHistoryStore`, not a concrete instance of `SqlChannelMemberHistoryStore`, so I can't call `GetMaster()` in it, because it isn't the concrete type, and I can't case the interface type to the concrete type because there could be multiple implementations of the interface, and the compiler won't know which one to use.

Reply:
```go
var channelMemberStore ChannelMemberHistoryStore
...

channelMemberSqlStore := channelMemberStore.(*SqlChannelMemberHistoryStore)
master := channelMemberSqlStore.GetMaster()
```

Reply:
> I can't case the interface type to the concrete type because there could be multiple implementations of the interface, and the compiler won't know which one to use.

Not really sure what you mean by this.

Other ways you could do it:

```go
channelMemberSqlStore, ok := channelMemberStore.(*SqlChannelMemberHistoryStore)
if !ok {
    // not using sql
    t.Skip()
}
```

```go
var master *gorp.DbMap
if sqlStore, ok := channelMemberStore.(interface {
    GetMaster() *gorp.DbMap
}); ok {
    master = sqlStore.GetMaster()
}

Reply:
Two things:
1. I don't understand how it's possible to cast an interface to a concrete type. The compiler shouldn't have enough information available to do that. What if the interface is some other type? Seems unsafe to me.
2. Adding that type assertion creates a circular dependency, because the `storetest` package now imports `sqlstore`, which in turn imports `storetest`. I'll try to unwind it, but I'm not sure that I can without doing a largish refactor.

Reply:
`storetest/channel_member_history_store.go` imports `store/sqlstore`, which loads `store/sqlstore/channel_member_history_store_test.go`, which completes the cycle by importing `store/storetest`. To fix this, we'd probably have to move the contents of the `storetest` package into `store/sqlstore`. That isn't ideal because we want to support other types of stores in the future. I think this prevents us from using a type assertion to explicitly get an instance of `SqlChannelMemberHistoryStore` in the test.

Reply:
Fixed in https://github.com/mattermost/mattermost-server/pull/7951/commits/830ad8a45e8f28b902d0e4c4025d4ee5b59e8da3